Scipy 1.15 release impacting `sph_harm` CI

Looks like there has been an upstream change on the OSX ARM platform following the Jan 3 release of scipy==1.15.0.

I'll try to patch this today so the CI clears up.

```
=========================== short test summary info ============================
FAILED tests/test_basis_utils.py::test_sph_harm_low_order - AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 1 / 42 (2.38%)
Max absolute difference among violations: 5.08309226e-48
Max relative difference among violations: 1.
 ACTUAL: array([ 0.      +0.j      , -0.001105-0.000547j, -0.005811-0.007624j,
       -0.005869-0.030263j,  0.01808 -0.065793j,  0.081064-0.090961j,
        0.174581-0.070362j,  0.260057+0.019966j,  0.284321+0.16903j ,...
 DESIRED: array([ 0.000000e+00+0.000000e+00j, -1.105183e-03-5.472108e-04j,
       -5.811352e-03-7.6[237](https://github.com/ComputationalCryoEM/ASPIRE-Python/actions/runs/12651714833/job/35254584553#step:5:238)76e-03j, -5.869114e-03-3.026275e-02j,
        1.808013e-02-6.579263e-02j,  8.106404e-02-9.096090e-02j,...
FAILED tests/test_basis_utils.py::test_sph_harm_high_order - AssertionError: assert not np.True_
 +  where np.True_ = <ufunc 'isfinite'>(np.complex128(-4.37681178857729e-80+5.07417939080972e-80j))
 +    where <ufunc 'isfinite'> = np.isfinite
= 2 failed, 5598 passed, 32 skipped, 82 xfailed, 61 warnings in 421.94s (0:07:01) =
Error: Process completed with exit code 1.
```

--- a/tests/test_basis_utils.py
+++ b/tests/test_basis_utils.py
@@ -1,7 +1,22 @@
 from unittest import TestCase
 
 import numpy as np
-from scipy.special import sph_harm_y
+
+# This can be removed when project requires scipy>=1.15.0
+# scipy<1.15.0 provide `sph_harm`
+import scipy
+from packaging.version import Version
+
+if Version(scipy.__version__) < Version("1.15.0"):
+    from scipy.special import sph_harm as sp_sph_harm
+
+    # This has a different convention from upstream sph_harm_y
+    def sph_harm_y(j, m, x, y):
+        return sp_sph_harm(m, j, y, x)
+
+else:
+    # scipy>=1.15.0 provide `sph_harm_y`
+    from scipy.special import sph_harm_y
 
 from aspire.basis.basis_utils import (
     all_besselj_zeros,

--- a/tests/test_basis_utils.py
+++ b/tests/test_basis_utils.py
@@ -1,7 +1,7 @@
 from unittest import TestCase
 
 import numpy as np
-from scipy.special import sph_harm as sp_sph_harm
+from scipy.special import sph_harm_y
 
 from aspire.basis.basis_utils import (
     all_besselj_zeros,
@@ -24,14 +24,14 @@ def test_sph_harm_low_order():
     x = np.linspace(0, np.pi, 42)
     y = np.linspace(0, 2 * np.pi, 42)
 
-    ref = sp_sph_harm(m, j, y, x)  # Note calling convention is different
+    ref = sph_harm_y(j, m, x, y)  # Note Scipy calling convention is different
     # Prescribe an atol because some of the ref values can be very
     # small, which can impact relative tolerance.
     np.testing.assert_allclose(sph_harm(j, m, x, y), ref, atol=1e-8)
 
     # negative m
     m *= -1
-    ref = sp_sph_harm(m, j, y, x)  # Note calling convention is different
+    ref = sph_harm_y(j, m, x, y)  # Note Scipy calling convention is different
     # Prescribe an atol because some of the ref values can be very
     # small, which can impact relative tolerance.
     np.testing.assert_allclose(sph_harm(j, m, x, y), ref, atol=1e-8)

--- a/tests/test_basis_utils.py
+++ b/tests/test_basis_utils.py
@@ -25,27 +25,33 @@ def test_sph_harm_low_order():
     y = np.linspace(0, 2 * np.pi, 42)
 
     ref = sp_sph_harm(m, j, y, x)  # Note calling convention is different
-    np.testing.assert_allclose(sph_harm(j, m, x, y), ref)
+    # Prescribe an atol because some of the ref values can be very
+    # small, which can impact relative tolerance.
+    np.testing.assert_allclose(sph_harm(j, m, x, y), ref, atol=1e-8)
 
     # negative m
     m *= -1
     ref = sp_sph_harm(m, j, y, x)  # Note calling convention is different
-    np.testing.assert_allclose(sph_harm(j, m, x, y), ref)
+    # Prescribe an atol because some of the ref values can be very
+    # small, which can impact relative tolerance.
+    np.testing.assert_allclose(sph_harm(j, m, x, y), ref, atol=1e-8)
 
 
 def test_sph_harm_high_order():
     """
-    Test we remain finite at higher orders where `scipy.special.sph_harm` overflows.
+    Test we remain finite at higher orders where legacy `scipy.special.sph_harm` overflowed.
     """
+    # Older (<1.15.0) versions of Scipy overflowed with these values.
+    # Scipy>=1.15.0 has better overflow behavior,
+    #   but the method `sph_harm` will be deprecated in 1.17.0.
+
     m = 87
     j = 87
     x = 0.12345
     y = 0.56789
 
-    # If scipy fixed their implementation for higher orders in the future,
-    # this check should fail and we can reconsider that package.
-    ref = sp_sph_harm(m, j, y, x)  # Note calling convention is different
-    assert not np.isfinite(ref)
+    # Check we are finite.
+    assert np.isfinite(sph_harm(j, m, x, y))
 
     # Can manually check against pyshtools,
     # but we are avoiding that package dependency.
@@ -60,9 +66,6 @@ def test_sph_harm_high_order():
     #     csphase=-1,
     #     normalization="ortho",
     # )
-
-    # Check we are finite.
-    assert np.isfinite(sph_harm(j, m, x, y))
 
 
 class BesselTestCase(TestCase):

